submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,300 @@
+## Globalnet 2.0: Enhancements to Globalnet Implementation
+
+## Summary
+
+Currently, Submariner Globalnet implementation in [release 0.8](https://github.com/submariner-io/releases/releases/tag/v0.8.0)
+annotates every Pod, Service and Node in the cluster with a globalIP to support Pod to remote
+Service connectivity in the following manner.
+
+Once a Pod is scheduled and enters Running state, it is annotated with a globalIP. To support
+the connectivity, Globalnet programs the necessary egress rules on the active Gateway Node of
+the cluster where the source IP of the Pod is modified to the corresponding globalIP. This
+translation enables us to uniquely identify the traffic from the Pod, when it reaches the
+destination cluster.
+
+Similarly, a ClusterIP Service is annotated with a globalIP so that remote Pods can communicate
+with it using its globalIP. Globalnet programs the necessary ingress rules on the active Gateway
+node of the cluster, where the service resides. The ingress rules enable any incoming traffic
+destined to the globalIP of the Service to be redirected to the corresponding ClusterIP Service.
+
+In order to support connectivity from the Host Network to the remote services, Globalnet also
+annotates every node in the cluster with a unique globalIP.
+
+While this works fine for small clusters, where the number of Pods and Services are limited, for
+large clusters it poses scalability and performance issues as we would eventually run out of
+globalIPs because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
+connectivity as well as Headless Services are not supported.
+
+## Proposal
+
+This enhancement proposes the following changes to make Globalnet implementation scale better.
+
+Egress traffic:
+
+* By default, every cluster will be assigned a configurable number of globalIPs which will be
+  used as egress IPs for any cross-cluster communication. Support for more than a single globalIP
+  at the cluster level is done to avoid ephemeral port exhaustion issues.
+* Provision to allocate a set of globalIPs at the namespace/project level which takes precedence
+  over the globalIPs allocated to the cluster. All Pods in the namespace will use the same globalIPs
+  as EgressIPs.
+* Provision to allocate globalIPs to a set of Pods in a selected namespace or cluster-wide.
+  Such Pods will use globalIPs that take precedence over any globalIPs allocated at the default
+  namespace level.
+
+Ingress traffic:
+
+* Instead of assigning a globalIP every ClusterIP Service in the cluster, only exported Services
+  will be assigned a globalIP. Globalnet internally creates a GlobalIngressIP CR for every
```

Comment:
Is it Okay if we discuss the CRD as part of implementation PR that adds support for Services? JFYI, the `GlobalIngressIP` is internally created by Globalnet when Service is exported and is not expected to be created by the users.

Reply:
> Is it Okay if we discuss the CRD as part of implementation PR that adds support for Services?

I think defining the CRD (even if internal) should help understand the ingress side of the proposal which is a bit vague IMHO, eg is it one IP or more? How is it connected to service? How is it designed for headless services?

> JFYI, the `GlobalIngressIP` is internally created by Globalnet when Service is exported and is not expected to be created by the users.

What would happen if a user creates one though?
Or would that be blocked by RBAC?

Reply:
> I think defining the CRD (even if internal) should help understand the ingress side of the proposal which is a bit vague IMHO, eg is it one IP or more? How is it connected to service? How is it designed for headless services?

I'd imagine that for every exported ClusterIP Service and Headless Service, we will internally create a GlobalnetIngress object in the same namespace with the same name. 

Regarding the number of IPs:
For an exported ClusterIP Service, there will only be a single globalIP as ingress IP.
For a Headless Service, for every backend pod, we will have a unique globalIP as ingress IP.

@vthapar please let me know if this works or if you think we will have to model it differently.

> What would happen if a user creates one though?
> Or would that be blocked by RBAC?

Ideally, I will expect it to be blocked by RBAC, but we will have more clarity when we start working on the implementation.

Reply:
Hence why I think it's best to define the CRD here just you're defining the other CRDs.

Another way of looking at it - why describe the other CRDs but not this one?

It's definitely an API and more a design detail than implementation detail.

Reply:
@mkolesnik @sridhargaddam I am leaning towards a GlobalIngress object for each pod in case of headless service instead of tying it to exported headless service. Way I see it, it is similar to user allocating/requesting global IP to a pod, just that in case of headless service we are doing it when a service is allocated. We can add labels to mark which services they belong to optimize lookups.

Reply:
If we define `GlobalIngress` object in a way that align to k8s `Service` and `Endpoints` object, it would be like:

<details>

```
type GlobalNetService struct {
	metav1.TypeMeta
	// +optional
	metav1.ObjectMeta

	// Spec defines the behavior of a service.
	// +optional
	Spec GlobalNetServiceSpec

	// Status represents the current status of a service.
	// +optional
	Status GlobalNetServiceStatus
}

type GlobalNetServiceSpec struct {
	// GlobalIP is the IP address of the service from outside a cluster
	// +optional
	GlobalIP string
}

type GlobalNetServiceStatus struct {
	// Current GlobalNetService condition
	// +optional
	Conditions []metav1.Condition
}

type GlobalNetEndpoints struct {
	metav1.TypeMeta
	// +optional
	metav1.ObjectMeta

	// The set of all endpoints is the union of all subsets.
	Subsets []GlobalNetEndpointSubset
}

type GlobalNetEndpointSubset struct {
	// GlobalIP is the IP address of the service from outside a cluster
	GlobalIP string

	// EndpointAddress is a tuple that describes single IP address.
	EndpointAddress v1.EndpointAddress

	Ports           []v1.EndpointPort
}
```

</details>

If these resources are created by controller when a `ServiceExport` is created by user and are reconciled on updates in `Service`, `Endpoints`, and `Pod`, it will cover ingress of exported `Service` for both normal and headless case. In addition, it will include definition of egress for headless service.

I'm also thinking about `GlobalEgress` can also be aligned to this (`GlobalEgress` will be just sets of `GlobalNetEndpointSubset` that has empty `Ports`, despite that there's no such concept in k8s, like `Service` without Ports).

Reply:
@mkimuram ServiceExport/Import are somewhat analogous to `GlobalNetService*` objects you're proposing. `GlobalnetIngress` object I was thinking of similar to what you've proposed as `GlobalnetEndpoints`